core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1754,6 +1754,7 @@ class UnifiedLog(@volatile var logStartOffset: Long,
         leaderEpochCache.foreach(_.clearAndFlush())
         producerStateManager.truncateFullyAndStartAt(newOffset)
         logStartOffset = logStartOffsetOpt.getOrElse(newOffset)
+        if (remoteLogEnabled()) maybeIncrementLocalLogStartOffset(newOffset, LogStartOffsetIncrementReason.SegmentDeletion)
```

Comment:
Is the reason segment deletion correct here? When an out of sync follower is catching up, it will truncate fully and start from beginning if the leader's local offset has moved past the followers local end offset. The leader's local offset could have moved past due to archival. 

I would suggest to look at cases where this function is called and perhaps pass the reason from those here if there are multiple different ones OR you can use LeaderOffsetIncremented as the reason.

Reply:
The reason is indeed not always correct. In the case of truncation when a follower aligns with the leader, what was missing was the update of the local log start offset, which was unchanged after building the state on the follower from auxiliary files and the response from the initial follower fetcher.

Reply:
Actually is this really the right place to update the local log start offset? The replica fetcher tier state machine already calls `maybeIncrementLogStartOffset` from [here](https://github.com/apache/kafka/blob/trunk/core/src/main/java/kafka/server/ReplicaFetcherTierStateMachine.java#L233). I guess we can add the call to `maybeIncrementLocalLogStartOffset` there too?

Reply:
I have made `maybeIncrementLocalLogStartOffset` public and calling it from `buildRemoteLogAuxState`.

Although I think that we should still update the local log start offset in `truncateFullyAndStartAt` because we know **exactly** what the start offset of the local log is here. The replica fetcher tier state machine already calls `maybeIncrementLogStartOffset`, but that call is redundant because we already update the log start offset in `truncateFullyAndStartAt` to the `logStartOffsetOpt` parameter (without specifying a reason).

Adding it to `truncateFullyAndStartAt` means that anything that calls the function does not have to worry about the updating the local log start offset every time. Although we should not hardcode the reason here because that will not be always correct. The two options then are (i) update directly like we do for the log start offset, or (ii) accept the reason as a parameter which means modify all callers, including lots of `truncateFully...` functions down the chain, and modify all the places including tests that call any of those functions.

I will leave this open to discussion. +cc @divijvaidya @kamalcph @satishd @Hangleton 

This PR in its current state is ready to be reviewed.

Reply:
> Adding it to truncateFullyAndStartAt means that anything that calls the function does not have to worry about the updating the local log start offset every time.

My preference is to update the local-log-start-offset in `truncateFullyAndStartAt` function as it updates the leader-epoch-checkpoint file, log-start-offset, high-watermark, and log-end-offset. 

To avoid passing the reason in maybeIncrementLocalLogStartOffset` method, we can update the `local-log-start-offset` value directly in truncateFullyAndStartAt method since it takes the same `lock`. 

We followed a similar approach in our local branch: https://sourcegraph.com/github.com/satishd/kafka@24d21694f2f0450e8a8e60e85400de3586fc7069/-/blob/core/src/main/scala/kafka/log/Log.scala?L2385

Reply:
Thanks @kamalcph. I agree with that. I have updated the PR to update the local log start offset in `truncateFullyAndStartAt` and also in  `buildRemoteLogAuxState` with the correct reason.